Sources/NIO/ControlMessage.swift

```diff
@@ -0,0 +1,245 @@
+    
+    struct Index: Equatable, Comparable {
+        fileprivate var cmsgPointer: UnsafeMutablePointer<cmsghdr>?
+        
+        static func < (lhs: UnsafeControlMessageCollection.Index,
+                       rhs: UnsafeControlMessageCollection.Index) -> Bool {
+            // nil is high, as that's the end of the collection.
+            switch (lhs.cmsgPointer, rhs.cmsgPointer) {
+            case (.some(let lhs), .some(let rhs)):
+                return lhs < rhs
+            case (.some, .none):
+                return true
+            case (.none, .some), (.none, .none):
+                return false
+            }
+        }
+        
+        fileprivate init(cmsgPointer: UnsafeMutablePointer<cmsghdr>?) {
+            self.cmsgPointer = cmsgPointer
+        }
+    }
+    
+    var startIndex: Index {
+        var messageHeader = self.messageHeader
+        return withUnsafePointer(to: &messageHeader) { messageHeaderPtr in
+            let firstCMsg = Posix.cmsgFirstHeader(inside: messageHeaderPtr)
+            return Index(cmsgPointer: firstCMsg)
+        }
+    }
+    
+    var endIndex: Index { return Index(cmsgPointer: nil) }
+    
+    func index(after: Index) -> Index {
+        precondition(after.cmsgPointer != nil)
+        var msgHdr = messageHeader
+        return withUnsafeMutablePointer(to: &msgHdr) { messageHeaderPtr in
+            return Index(cmsgPointer: Posix.cmsgNextHeader(inside: messageHeaderPtr,
+                                                           after: after.cmsgPointer!))
+        }
+    }
+    
+    public subscript(position: Index) -> Element {
+        let cmsg = position.cmsgPointer!
+        return UnsafeControlMessage(level: cmsg.pointee.cmsg_level,
+                                    type: cmsg.pointee.cmsg_type,
+                                    data: Posix.cmsgData(for: cmsg))
+    }
+}
+
+/// Extract information from a collection of control messages.
+struct ControlMessageParser {
+    var ecnValue: NIOExplicitCongestionNotificationState = .transportNotCapable // Default
+    
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    private static let ipv4TosType = IP_RECVTOS
+    #else
+    private static let ipv4TosType = IP_TOS    // Linux
+    #endif
+
+    static func readCInt(data: UnsafeRawBufferPointer) -> CInt {
+        assert(data.count == MemoryLayout<CInt>.size)
+        precondition(data.count >= MemoryLayout<CInt>.size)
+        var readValue = CInt(0)
+        withUnsafeMutableBytes(of: &readValue) { valuePtr in
+            valuePtr.copyMemory(from: data)
+        }
+        return readValue
+    }
+    
+    mutating func receiveMessage(_ controlMessage: UnsafeControlMessage) {
+        if controlMessage.level == IPPROTO_IP && controlMessage.type == ControlMessageParser.ipv4TosType {
+            if let data = controlMessage.data {
+                assert(data.count == 1)
+                precondition(data.count >= 1)
```

Comment:
Seems like at least one of these is wrong/unnecessary!

Reply:
Nope, they're doing different things.

The `assert` codifies something we believe to be true, but that is fundamentally outside our control. The `precondition` covers something that we _require_ to be true for safety. We _believe_ we'll only ever have one byte of data to read from, but it doesn't actually matter if we have more: we're just going to ignore the others. So we use `assert` to guarantee our programming expectations, but at runtime we tolerate machines that do something different so long as they don't lead to us doing something unsafe.

Reply:
Oh I see now, thanks for the explanation.